Isolate _StridedShard from Shard

zpcore opened the pull request:
### Summary

Separate out `_StridedShard` class from `Shard`.

### Details
1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Supports Redistribution for any StridedShard placement that is convertible to ordered Shard placement
5. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
    1. Operations impacted: operation that works on model parameters. From what I have seen, those operations are pointwise foreach operations. E.g., `optimizer.step()`, `torch.nn.utils.clip_grad_norm_()`, `GradScaler.step()`. 
6. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.
    1. view ops: fix `maybe_get_shard_mesh_dim_and_placement()` to propagate StridedShard.


Note: If we isolate StridedShard from Shard, generally this is what will happen if we don't change sharding strategy to support StridedShard: https://github.com/pytorch/pytorch/issues/166598#issuecomment-3583225318

### Analysis of impact to pointwise ops
**TL;DR** `StridedShard` should be propagated though with the same split_factor. Correctness not impacted.

The change is centralized in `map_placements_after_reduction`, where the function handles two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.

### Analysis of impact to Optimizer in FSDP:
**TL;DR** `StridedShard` can pass though foreach pointwise ops in the optimizer related to the model parameters, and will not impact the correctness of FSDP, no SS->S or S->SS redistribution happen.

I added a test `Test_StridedShard_Optimizer` in this PR. The test verified that different optimizers updated the grad for the fsdp model (contains `StridedShard` in parameters) correctly compared with the unsharded model. I also checked that for different optimizers, what aten ops will take StridedShard in inputs' placements as in the following table. I see three cases for those optimizers:
1. `detach`, `zeros_like`, `full_like`, clone all passes SS through input to output. I guess this happens in model weight initialization.
2. out-of-place foreach, like `_foreach_add.Scalar`, the SS will pass through to output spec.
3. in-place foreach. Those ops are kind of weird that they don't have output spec (If we check the output spec, it's None. I debugged and noticed that this is due to the metadata propagation for fake tensor is None when we call inplace foreach op.). However, I can see it generates new input spec for each arg. Those new input spec will still keep SS if the original input spec is SS.

So I am kind of confident now that this PR allows StridedShard pass though pointwise ops in the optimizer and will not impact the correctness of FSDP. 

- reference: modified [_sharding_prop.py](https://gist.github.com/zpcore/fc81a0017fd673333ac282b67c750bf2) for testing what ops received inputs that contain StridedShard and whether StridedShard has passed through aten op or not.

| Optimizer      | Aten Op                        |  
|----------------|-------------------------------|
| adadelta       | `detach.default`              |
|                | `zeros_like.default`         |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_add.Scalar`          |
|                | `_foreach_sqrt_.default`       |
|                | `_foreach_div_.List`           |
|                | `_foreach_mul_.List`           |
|                | `_foreach_add_.List`           |
| adagrad        | `detach.default`         |
|                | `full_like.default`          |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`       |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_mul.ScalarList`    |
|                | `_foreach_addcdiv_.Scalar`     |
| adam           | `detach.default`         |
|                | `zeros_like.default`       |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`     |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList ` |
| adamw          | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList` |
| rmsprop        | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.Scalar`     |
| sgd            | `detach.default`               |
|                | `_foreach_add_.List`           |
| sgd (momentum) | `detach.default`               |
|                | `clone.default`                |
|                | `_foreach_add_.List`           |
|                | `_foreach_add_.List`           |
|                | `_foreach_mul_.Scalar`        |


### Analysis of impact to view ops
**TL;DR** Previous and current code can propagate `StridedShard` by maintaining the same split_factor (which I don't think this is correct for all cases), as long as the tensor dim is shardable on the mesh. We need follow up fix the let view op propagate arbitrary combination of `StridedShard`.

I marked `expectedFailure` for `test_view_propagation_not_supported_yet` in this PR. @weifengpy plans to work on a broader support.

### Analysis of impact to random ops
Our current `OffsetBasedRNGTracker` uses the Shard placement to compute the shard idx of each shard the shard size. This will be used to decide the RNG offset. However, the offset computation didn't consider the split_factor in the `StridedShard`.

This PR didn't fix any offset computation based on `StridedShard`. It only allows the `StridedShard` to be treated equally as Shard, which is the same behavior before this PR when `StridedShard` is a subclass of Shard.


---
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167300
* __->__ #167887



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

zpcore committed:
[Experimental] Make _StridedShard not a subclass of Shard

[ghstack-poisoned]

zpcore committed:
Update on "[Experimental] Make _StridedShard not a subclass of Shard"

[Not for Review yet] Separate out _StridedShard class from Shard.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[Experimental] Make _StridedShard not a subclass of Shard"

[Not for Review yet] Separate out _StridedShard class from Shard.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[Experimental] Make _StridedShard not a subclass of Shard"

[Not for Review yet] Separate out _StridedShard class from Shard.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/_C/_distributed.pyi:
```diff
@@ -10,6 +10,7 @@ class Shard(Placement):
     def __init__(self, dim: int): ...
 
 class StridedShard(Shard):
```
Update this too

ezyang reviewed (commented):
Looks reasonable. Please be noisy about problems

zpcore committed:
Update on "[Experimental] Make _StridedShard not a subclass of Shard"

[Not for Review yet] Separate out _StridedShard class from Shard.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Make _StridedShard not a subclass of Shard"

Separate out _StridedShard class from Shard.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Make _StridedShard not a subclass of Shard"

Separate out _StridedShard class from Shard.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

Separate out `_StridedShard` class from `Shard`.

1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
5. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented:
This seems to indicate that we need to support redistribute on strided shard explicitly first. Which... seems reasonable.

zpcore committed:
Update on "Isolate _StridedShard from Shard"

Separate out `_StridedShard` class from `Shard`.

1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
5. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

Separate out `_StridedShard` class from `Shard`.

1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
5. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

Separate out `_StridedShard` class from `Shard`.

1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
5. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

Separate out `_StridedShard` class from `Shard`.

1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
5. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/Placement.h:
```diff
@@ -53,27 +56,18 @@ class Shard : public Placement {
   std::int64_t split_factor;
-  explicit StridedShard(std::int64_t dim, std::int64_t split_factor_)
-      : Shard(dim), split_factor(split_factor_) {}
+  explicit StridedShard(std::int64_t dim_, std::int64_t split_factor_)
```
nit: why do we add underscore to `dim_` here? was it intentional in `Shard` or was that just a typo?

wconstab commented on torch/distributed/tensor/_api.py:
```diff
@@ -800,23 +809,25 @@ def distribute_tensor(
+            )
+            placements[idx] = Shard(placement_dim)
+        elif isinstance(placement, _StridedShard):
+            placement_dim = (
```
nit: move this to a common block above?

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -93,9 +93,37 @@ def __post_init__(self) -> None:
             # pyrefly: ignore [bad-assignment]
-            self.shard_order = DTensorSpec.compute_default_shard_order(self.placements)
+            if any(isinstance(p, _StridedShard) for p in self.placements):
+                self.shard_order = self._maybe_convert_StridedShard_to_shard_order(
```
what should happen here if `_maybe_convert` refuses to convert?

wconstab replied:
should we just replace this if/else block with `self.shard_order = _normalize_placements...`?  
and add a comment explaining that if `self.shard_order` is None, that means the placements contains un-normalizable SS?

zpcore commented on torch/distributed/tensor/_ops/_math_ops.py:
```diff
@@ -267,7 +268,14 @@ def map_placements_after_reduction(
+                    new_placements.append(Shard(new_shard_dim))
+                else:
+                    new_placements.append(
+                        _StridedShard(
```
I added a simple test `test_pointwise_propagation`. So the `map_placements_after_reduction` will handle two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.

zpcore commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -574,6 +578,56 @@ def _make_shard_tensor(
+        mesh_dim: int,
+        src_data_rank: Optional[int] = 0,
+    ) -> torch.Tensor:
+        """
```
I didn't modify the logic inside `_shard_tensor`. Basically I copy the function here from Shard and let this _shard_tensor calls SS's _split_tensor method, because SS is not a subclass of Shard anymore.

zpcore committed:
Update on "Isolate _StridedShard from Shard"

### Summary

Separate out `_StridedShard` class from `Shard`.

### Details
1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Supports Redistribution for any StridedShard placement that is convertible to ordered Shard placement
5. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
    1. Operations impacted: operation that works on model parameters. From what I have seen, those operations are pointwise foreach operations. E.g., `optimizer.step()`, `torch.nn.utils.clip_grad_norm_()`, `GradScaler.step()`. 
6. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.
    1. view ops: fix `maybe_get_shard_mesh_dim_and_placement()` to propagate StridedShard.


Note: If we isolate StridedShard from Shard, generally this is what will happen if we don't change sharding strategy to support StridedShard: https://github.com/pytorch/pytorch/issues/166598#issuecomment-3583225318


### Analysis of impact to pointwise ops
**TL;DR** `StridedShard` should be propagated though with the same split_factor. Correctness not impacted.

The change is centralized in `map_placements_after_reduction`, where the function handles two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.


### Analysis of impact to view ops
**TL;DR** Current code can correctly propagate StridedShard that didn't get flattened/unflattened from view op. 

First, let's see how StridedShard is expected to work with view sharding propagation. Below table shows examples on how shard is propagated through view ops over device mesh shape [2, 2, 2, 2]:
(copied from [internal slide](https://docs.google.com/presentation/d/1zYNotExnd4lG-B1eijX6CSHM6jYKWRlvMcF5TB-MbPs/edit?slide=id.g398133a0f1b_0_28#slide=id.g398133a0f1b_0_28))

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-ef1ce3e7-7fff-d4c6-627b-aef3d69070f7">
origin tensor → viewed tensor | sharding on original tensor | *Equivalent sharding representation on viewed tensor
-- | -- | --
[8, 8] → [2, 4, 8] | [S(0), R, R, R] | [S(0), R, R, R]
[8, 8] → [2, 4, 8] | [S(1), R, R, R] | [S(2), R, R, R]
[4, 4, 8] → [4, 4, 2, 4] | [ _SS(0, sf=2), S(2), R, R] | [ _SS(0, sf=2), S(2), R, R]
[4, 4, 8] → [4, 4, 2, 4] | [_SS(2, sf=2), R, R, R] | [S(3), R, R, R]
[4, 4, 8] → [4, 4, 2, 4] | [_SS(2, sf=2), S(2), R, R] | [S(3), S(2), R, R]
[4, 4, 8] → [4, 4, 2, 4] | [S(2), _SS(2, sf=2), R, _SS(0, sf=2)] | [S(2), _SS(3, 2), R, _SS(0, sf=2)]
[4, 4, 8] → [4, 4, 2, 4] | [_SS(2, sf=4), S(2), R, R] | [_SS(3, sf=2), S(2), R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(0), R, R, R] | [S(0), R, R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(1), R, R, R] | [ _SS(0, sf=ts[0]), R, R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(2), R, R, R] | [S(1), R, R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(0), S(1), R, R] | [S(0), S(0), R, R]
[4, 4, 8] → [16, 8] | [S(0), S(0), R, R] | [S(0), S(0), R, R]
[4, 4, 8] → [16, 8] | [S(1), S(1), R, R] | [_SS(0, sf=ts[0]), _SS(0, sf=ts[0]), R, R]
[4, 4, 8] → [16, 8] | [S(1), S(1), R, R] (right-to-left-order) | [_SS(0, sf=ts[0]*ms[1]), _SS(0, sf=ts[0]), R, R]
[2, 4, 4, 2] → [2, 16, 2] | [S(1), R, R, R] | [S(1), R, R, R]
[2, 4, 4, 2] → [2, 16, 2] | [S(2), R, R, R] | [_SS(1, sf=ts[1]), R, R, R]

</b>

\* ts[i]: tensor shape at i before view, ms: mesh shape

Note in row 2, 3, 4, 5, 12, 13, we should support flatten/unflatten a StridedShard tensor dim. This will be done in future PRs.

### Analysis of impact to random ops











---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci 

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

### Summary

Separate out `_StridedShard` class from `Shard`.

### Details
1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Supports Redistribution for any StridedShard placement that is convertible to ordered Shard placement
5. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
    1. Operations impacted: operation that works on model parameters. From what I have seen, those operations are pointwise foreach operations. E.g., `optimizer.step()`, `torch.nn.utils.clip_grad_norm_()`, `GradScaler.step()`. 
6. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.
    1. view ops: fix `maybe_get_shard_mesh_dim_and_placement()` to propagate StridedShard.


Note: If we isolate StridedShard from Shard, generally this is what will happen if we don't change sharding strategy to support StridedShard: https://github.com/pytorch/pytorch/issues/166598#issuecomment-3583225318

### Analysis of impact to pointwise ops
**TL;DR** `StridedShard` should be propagated though with the same split_factor. Correctness not impacted.

The change is centralized in `map_placements_after_reduction`, where the function handles two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.

### Analysis of impact to Optimizer in FSDP:
**TL;DR** `StridedShard` can pass though foreach pointwise ops in the optimizer related to the model parameters, and will not impact the correctness of FSDP. 

I added a test `Test_StridedShard_Optimizer` in this PR. The test verified that different optimizers updated the grad for the fsdp model (contains 1StridedShard1 in parameters) correctly compared with the unsharded model. I also checked that for different optimizers, what aten ops will take StridedShard in inputs' placements as in the following table. As we can see most are inplace `foreach` ops. For those ops that are not inplace, we can see StridedShard can be propagated through. So I am kind of confident now that this PR allows StridedShard pass though pointwise ops in the optimizer and will not impact the correctness of FSDP. 

- reference: modified [_sharding_prop.py](https://gist.github.com/zpcore/fc81a0017fd673333ac282b67c750bf2) for testing what ops received inputs that contain StridedShard and whether StridedShard has passed through aten op or not.

| Optimizer      | Aten Op                        |    Did StridedShard passed through? |
|----------------|-------------------------------|-----|
| adadelta       | `detach.default`              |    Y |
|                | `zeros_like.default`         | Y |
|                | `_foreach_mul_.Scalar`         | |
|                | `_foreach_addcmul_.Scalar`   |  |
|                | `_foreach_add.Scalar`          | |
|                | `_foreach_sqrt_.default`       | |
|                | `_foreach_div_.List`           | |
|                | `_foreach_mul_.List`           | |
|                | `_foreach_add_.List`           | |
| adagrad        | `detach.default`         | Y |
|                | `full_like.default`          |  Y |
|                | `_foreach_addcmul_.Scalar`     | |
|                | `_foreach_sqrt.default`       | Y |
|                | `_foreach_add_.Scalar`         | |
|                | `_foreach_mul.ScalarList`    |  Y |
|                | `_foreach_addcdiv_.Scalar`     | |
| adam           | `detach.default`         |   Y |
|                | `zeros_like.default`       |  Y |
|                | `_foreach_lerp_.Scalar`        | |
|                | `_foreach_mul_.Scalar`         | |
|                | `_foreach_addcmul_.Scalar`     | |
|                | `_foreach_sqrt.default`     |  Y |
|                | `_foreach_div_.ScalarList`     | |
|                | `_foreach_add_.Scalar`         | |
|                | `_foreach_addcdiv_.ScalarList ` | |
| adamw          | `detach.default`               | Y |
|                | `zeros_like.default`           | Y |
|                | `_foreach_mul_.Scalar`         | |
|                | `_foreach_lerp_.Scalar`        | |
|                | `_foreach_addcmul_.Scalar`   |  |
|                | `_foreach_sqrt.default`        |  Y |
|                | `_foreach_div_.ScalarList`     |  |
|                | `_foreach_add_.Scalar`         |  |
|                | `_foreach_addcdiv_.ScalarList` |  |
| rmsprop        | `detach.default`               |  Y |
|                | `zeros_like.default`           | Y |
|                | `_foreach_mul_.Scalar`         |  |
|                | `_foreach_addcmul_.Scalar`     | |
|                | `_foreach_sqrt.default`        | Y |
|                | `_foreach_add_.Scalar`         | |
|                | `_foreach_addcdiv_.Scalar`     | |
| sgd            | `detach.default`               | Y |
|                | `_foreach_add_.List`           | |
| sgd (momentum) | `detach.default`               |  Y |
|                | `clone.default`                | Y |
|                | `_foreach_add_.List`           | |
|                | `_foreach_add_.List`           | |
|                | `_foreach_mul_.Scalar`        |  |


### Analysis of impact to view ops
**TL;DR** Current code can correctly propagate StridedShard that didn't get flattened/unflattened from view op. 

First, let's see how StridedShard is expected to work with view sharding propagation. Below table shows examples on how shard is propagated through view ops over device mesh shape [2, 2, 2, 2]:
(copied from [internal slide](https://docs.google.com/presentation/d/1zYNotExnd4lG-B1eijX6CSHM6jYKWRlvMcF5TB-MbPs/edit?slide=id.g398133a0f1b_0_28#slide=id.g398133a0f1b_0_28))

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-ef1ce3e7-7fff-d4c6-627b-aef3d69070f7">
origin tensor → viewed tensor | sharding on original tensor | *Equivalent sharding representation on viewed tensor
-- | -- | --
[8, 8] → [2, 4, 8] | [S(0), R, R, R] | [S(0), R, R, R]
[8, 8] → [2, 4, 8] | [S(1), R, R, R] | [S(2), R, R, R]
[4, 4, 8] → [4, 4, 2, 4] | [ _SS(0, sf=2), S(2), R, R] | [ _SS(0, sf=2), S(2), R, R]
[4, 4, 8] → [4, 4, 2, 4] | [_SS(2, sf=2), R, R, R] | [S(3), R, R, R]
[4, 4, 8] → [4, 4, 2, 4] | [_SS(2, sf=2), S(2), R, R] | [S(3), S(2), R, R]
[4, 4, 8] → [4, 4, 2, 4] | [S(2), _SS(2, sf=2), R, _SS(0, sf=2)] | [S(2), _SS(3, 2), R, _SS(0, sf=2)]
[4, 4, 8] → [4, 4, 2, 4] | [_SS(2, sf=4), S(2), R, R] | [_SS(3, sf=2), S(2), R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(0), R, R, R] | [S(0), R, R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(1), R, R, R] | [ _SS(0, sf=ts[0]), R, R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(2), R, R, R] | [S(1), R, R, R]
[2, 4, 4, 2] → [8, 4, 2] | [S(0), S(1), R, R] | [S(0), S(0), R, R]
[4, 4, 8] → [16, 8] | [S(0), S(0), R, R] | [S(0), S(0), R, R]
[4, 4, 8] → [16, 8] | [S(1), S(1), R, R] | [_SS(0, sf=ts[0]), _SS(0, sf=ts[0]), R, R]
[4, 4, 8] → [16, 8] | [S(1), S(1), R, R] (right-to-left-order) | [_SS(0, sf=ts[0]*ms[1]), _SS(0, sf=ts[0]), R, R]
[2, 4, 4, 2] → [2, 16, 2] | [S(1), R, R, R] | [S(1), R, R, R]
[2, 4, 4, 2] → [2, 16, 2] | [S(2), R, R, R] | [_SS(1, sf=ts[1]), R, R, R]

</b>

\* ts[i]: tensor shape at i before view, ms: mesh shape

Note in row 2, 3, 4, 5, 12, 13, we should support flatten/unflatten a StridedShard tensor dim. This will be done in future PRs.

### Analysis of impact to random ops











---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci 

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

### Summary

Separate out `_StridedShard` class from `Shard`.

### Details
1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Supports Redistribution for any StridedShard placement that is convertible to ordered Shard placement
5. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
    1. Operations impacted: operation that works on model parameters. From what I have seen, those operations are pointwise foreach operations. E.g., `optimizer.step()`, `torch.nn.utils.clip_grad_norm_()`, `GradScaler.step()`. 
6. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.
    1. view ops: fix `maybe_get_shard_mesh_dim_and_placement()` to propagate StridedShard.


Note: If we isolate StridedShard from Shard, generally this is what will happen if we don't change sharding strategy to support StridedShard: https://github.com/pytorch/pytorch/issues/166598#issuecomment-3583225318

### Analysis of impact to pointwise ops
**TL;DR** `StridedShard` should be propagated though with the same split_factor. Correctness not impacted.

The change is centralized in `map_placements_after_reduction`, where the function handles two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.

### Analysis of impact to Optimizer in FSDP:
**TL;DR** `StridedShard` can pass though foreach pointwise ops in the optimizer related to the model parameters, and will not impact the correctness of FSDP, no SS->S or S->SS redistribution happen.

I added a test `Test_StridedShard_Optimizer` in this PR. The test verified that different optimizers updated the grad for the fsdp model (contains `StridedShard` in parameters) correctly compared with the unsharded model. I also checked that for different optimizers, what aten ops will take StridedShard in inputs' placements as in the following table. I see three cases for those optimizers:
1. `detach`, `zeros_like`, `full_like`, clone all passes SS through input to output. I guess this happens in model weight initialization.
2. out-of-place foreach, like `_foreach_add.Scalar`, the SS will pass through to output spec.
3. in-place foreach. Those ops are kind of weird that they don't have output spec (If we check the output spec, it's None. I debugged and noticed that this is due to the metadata propagation for fake tensor is None when we call inplace foreach op.). However, I can see it generates new input spec for each arg. Those new input spec will still keep SS if the original input spec is SS.

So I am kind of confident now that this PR allows StridedShard pass though pointwise ops in the optimizer and will not impact the correctness of FSDP. 

- reference: modified [_sharding_prop.py](https://gist.github.com/zpcore/fc81a0017fd673333ac282b67c750bf2) for testing what ops received inputs that contain StridedShard and whether StridedShard has passed through aten op or not.

| Optimizer      | Aten Op                        |  
|----------------|-------------------------------|
| adadelta       | `detach.default`              |
|                | `zeros_like.default`         |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_add.Scalar`          |
|                | `_foreach_sqrt_.default`       |
|                | `_foreach_div_.List`           |
|                | `_foreach_mul_.List`           |
|                | `_foreach_add_.List`           |
| adagrad        | `detach.default`         |
|                | `full_like.default`          |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`       |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_mul.ScalarList`    |
|                | `_foreach_addcdiv_.Scalar`     |
| adam           | `detach.default`         |
|                | `zeros_like.default`       |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`     |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList ` |
| adamw          | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList` |
| rmsprop        | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.Scalar`     |
| sgd            | `detach.default`               |
|                | `_foreach_add_.List`           |
| sgd (momentum) | `detach.default`               |
|                | `clone.default`                |
|                | `_foreach_add_.List`           |
|                | `_foreach_add_.List`           |
|                | `_foreach_mul_.Scalar`        |


### Analysis of impact to view ops
**TL;DR** Previous and current code can propagate `StridedShard` by maintaining the same split_factor (which I don't think this is correct for all cases), as long as the tensor dim is shardable on the mesh. We need follow up fix the let view op propagate arbitrary combination of `StridedShard`.

I marked `expectedFailure` for `test_view_propagation_not_supported_yet` in this PR. weifengpy plans to work on a broader support.

### Analysis of impact to random ops
Our current `OffsetBasedRNGTracker` uses the Shard placement to compute the shard idx of each shard the shard size. This will be used to decide the RNG offset. However, the offset computation didn't consider the split_factor in the `StridedShard`.

This PR didn't fix any offset computation based on `StridedShard`. It only allows the `StridedShard` to be treated equally as Shard, which is the same behavior before this PR when `StridedShard` is a subclass of Shard.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci 

[ghstack-poisoned]

zpcore commented on test/distributed/tensor/test_utils.py:
```diff
@@ -994,6 +1003,283 @@ def test_2d_mesh_uneven_strided_shard(self):
+            assert isinstance(res1, DTensor)
+            assert isinstance(res2, DTensor)
+            self.assertEqual(res1.full_tensor(), res2.full_tensor())
+            self.assertEqual(res1.full_tensor(), res2.full_tensor())
```
Removed

zpcore commented on test/distributed/tensor/test_utils.py:
```diff
@@ -994,6 +1003,283 @@ def test_2d_mesh_uneven_strided_shard(self):
+            res1 = torch.sum(A, dim=0)
+            assert isinstance(res1, DTensor)
+            self.assertEqual(res1.full_tensor(), torch.sum(input_tensor, dim=0))
+            res2 = torch.sum(A, dim=1)
```
Added, yes, there should be no comm happening.

zpcore commented on test/distributed/tensor/test_utils.py:
```diff
@@ -994,6 +1003,283 @@ def test_2d_mesh_uneven_strided_shard(self):
+                input_tensor, mesh, [_StridedShard(0, split_factor=4), Shard(0)]
+            )
+            res1 = A @ B1
+            res2 = A @ B2
```
Added the commdebug. `res2 = A @ B2` will trigger 3 allgather due to redistribution.

zpcore committed:
Update on "Isolate _StridedShard from Shard"

### Summary

Separate out `_StridedShard` class from `Shard`.

### Details
1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Supports Redistribution for any StridedShard placement that is convertible to ordered Shard placement
5. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
    1. Operations impacted: operation that works on model parameters. From what I have seen, those operations are pointwise foreach operations. E.g., `optimizer.step()`, `torch.nn.utils.clip_grad_norm_()`, `GradScaler.step()`. 
6. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.
    1. view ops: fix `maybe_get_shard_mesh_dim_and_placement()` to propagate StridedShard.


Note: If we isolate StridedShard from Shard, generally this is what will happen if we don't change sharding strategy to support StridedShard: https://github.com/pytorch/pytorch/issues/166598#issuecomment-3583225318

### Analysis of impact to pointwise ops
**TL;DR** `StridedShard` should be propagated though with the same split_factor. Correctness not impacted.

The change is centralized in `map_placements_after_reduction`, where the function handles two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.

### Analysis of impact to Optimizer in FSDP:
**TL;DR** `StridedShard` can pass though foreach pointwise ops in the optimizer related to the model parameters, and will not impact the correctness of FSDP, no SS->S or S->SS redistribution happen.

I added a test `Test_StridedShard_Optimizer` in this PR. The test verified that different optimizers updated the grad for the fsdp model (contains `StridedShard` in parameters) correctly compared with the unsharded model. I also checked that for different optimizers, what aten ops will take StridedShard in inputs' placements as in the following table. I see three cases for those optimizers:
1. `detach`, `zeros_like`, `full_like`, clone all passes SS through input to output. I guess this happens in model weight initialization.
2. out-of-place foreach, like `_foreach_add.Scalar`, the SS will pass through to output spec.
3. in-place foreach. Those ops are kind of weird that they don't have output spec (If we check the output spec, it's None. I debugged and noticed that this is due to the metadata propagation for fake tensor is None when we call inplace foreach op.). However, I can see it generates new input spec for each arg. Those new input spec will still keep SS if the original input spec is SS.

So I am kind of confident now that this PR allows StridedShard pass though pointwise ops in the optimizer and will not impact the correctness of FSDP. 

- reference: modified [_sharding_prop.py](https://gist.github.com/zpcore/fc81a0017fd673333ac282b67c750bf2) for testing what ops received inputs that contain StridedShard and whether StridedShard has passed through aten op or not.

| Optimizer      | Aten Op                        |  
|----------------|-------------------------------|
| adadelta       | `detach.default`              |
|                | `zeros_like.default`         |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_add.Scalar`          |
|                | `_foreach_sqrt_.default`       |
|                | `_foreach_div_.List`           |
|                | `_foreach_mul_.List`           |
|                | `_foreach_add_.List`           |
| adagrad        | `detach.default`         |
|                | `full_like.default`          |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`       |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_mul.ScalarList`    |
|                | `_foreach_addcdiv_.Scalar`     |
| adam           | `detach.default`         |
|                | `zeros_like.default`       |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`     |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList ` |
| adamw          | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList` |
| rmsprop        | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.Scalar`     |
| sgd            | `detach.default`               |
|                | `_foreach_add_.List`           |
| sgd (momentum) | `detach.default`               |
|                | `clone.default`                |
|                | `_foreach_add_.List`           |
|                | `_foreach_add_.List`           |
|                | `_foreach_mul_.Scalar`        |


### Analysis of impact to view ops
**TL;DR** Previous and current code can propagate `StridedShard` by maintaining the same split_factor (which I don't think this is correct for all cases), as long as the tensor dim is shardable on the mesh. We need follow up fix the let view op propagate arbitrary combination of `StridedShard`.

I marked `expectedFailure` for `test_view_propagation_not_supported_yet` in this PR. weifengpy plans to work on a broader support.

### Analysis of impact to random ops
Our current `OffsetBasedRNGTracker` uses the Shard placement to compute the shard idx of each shard the shard size. This will be used to decide the RNG offset. However, the offset computation didn't consider the split_factor in the `StridedShard`.

This PR didn't fix any offset computation based on `StridedShard`. It only allows the `StridedShard` to be treated equally as Shard, which is the same behavior before this PR when `StridedShard` is a subclass of Shard.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci 

[ghstack-poisoned]

zpcore committed:
Update on "Isolate _StridedShard from Shard"

### Summary

Separate out `_StridedShard` class from `Shard`.

### Details
1. `_StridedShard(dim, sf=...)==Shard(dim)` should return False.
2. `is_shard()` should only be true for `Shard` class.
3. _StridedShard becomes not a subclass of Shard
4. Supports Redistribution for any StridedShard placement that is convertible to ordered Shard placement
5. Fix pointwise related propagation to support `_StridedShard`. Originally `_StridedShard` was silently cast into `Shard`. This will cause FSDP optimizer to update the model parameters incorrectly.
    1. Operations impacted: operation that works on model parameters. From what I have seen, those operations are pointwise foreach operations. E.g., `optimizer.step()`, `torch.nn.utils.clip_grad_norm_()`, `GradScaler.step()`. 
6. Some minor fixes to op sharding propagation regarding `_StridedShard` type, which impacts ops like rand.
    1. view ops: fix `maybe_get_shard_mesh_dim_and_placement()` to propagate StridedShard.


Note: If we isolate StridedShard from Shard, generally this is what will happen if we don't change sharding strategy to support StridedShard: https://github.com/pytorch/pytorch/issues/166598#issuecomment-3583225318

### Analysis of impact to pointwise ops
**TL;DR** `StridedShard` should be propagated though with the same split_factor. Correctness not impacted.

The change is centralized in `map_placements_after_reduction`, where the function handles two cases:
1. If reduction dims contains StridedShard, it replaces the reduction dims in original placements from Shard/StridedShard into Partial(...). E.g., for torch.sum(dim=0), we have strategy `SS(0)S(0)SS(0) --> PPP`. It should work because:
    - For all those Partial type like [sum, avg, product, max, min], they are independent of the whether it's stridedshard or normal shard. E.g. in partial('sum'), we sum up all dim that uses Shard(0) and StridedShard(0). Then it should work for arbitraty StridedShard combinations regardless of split factor and ordering.
2. For non reduction dims that contains StridedShard, e.g., `S(1)SS(1)`, we keep the original sharding but with an adjusted shard dim `S(1)SS(1)->S(0)SS(0)` when we reduce tensor dim 0. This should also work fine because we didn't touch any of the sharding on tensor dim 1.

### Analysis of impact to Optimizer in FSDP:
**TL;DR** `StridedShard` can pass though foreach pointwise ops in the optimizer related to the model parameters, and will not impact the correctness of FSDP, no SS->S or S->SS redistribution happen.

I added a test `Test_StridedShard_Optimizer` in this PR. The test verified that different optimizers updated the grad for the fsdp model (contains `StridedShard` in parameters) correctly compared with the unsharded model. I also checked that for different optimizers, what aten ops will take StridedShard in inputs' placements as in the following table. I see three cases for those optimizers:
1. `detach`, `zeros_like`, `full_like`, clone all passes SS through input to output. I guess this happens in model weight initialization.
2. out-of-place foreach, like `_foreach_add.Scalar`, the SS will pass through to output spec.
3. in-place foreach. Those ops are kind of weird that they don't have output spec (If we check the output spec, it's None. I debugged and noticed that this is due to the metadata propagation for fake tensor is None when we call inplace foreach op.). However, I can see it generates new input spec for each arg. Those new input spec will still keep SS if the original input spec is SS.

So I am kind of confident now that this PR allows StridedShard pass though pointwise ops in the optimizer and will not impact the correctness of FSDP. 

- reference: modified [_sharding_prop.py](https://gist.github.com/zpcore/fc81a0017fd673333ac282b67c750bf2) for testing what ops received inputs that contain StridedShard and whether StridedShard has passed through aten op or not.

| Optimizer      | Aten Op                        |  
|----------------|-------------------------------|
| adadelta       | `detach.default`              |
|                | `zeros_like.default`         |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_add.Scalar`          |
|                | `_foreach_sqrt_.default`       |
|                | `_foreach_div_.List`           |
|                | `_foreach_mul_.List`           |
|                | `_foreach_add_.List`           |
| adagrad        | `detach.default`         |
|                | `full_like.default`          |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`       |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_mul.ScalarList`    |
|                | `_foreach_addcdiv_.Scalar`     |
| adam           | `detach.default`         |
|                | `zeros_like.default`       |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`     |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList ` |
| adamw          | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_lerp_.Scalar`        |
|                | `_foreach_addcmul_.Scalar`   |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_div_.ScalarList`     |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.ScalarList` |
| rmsprop        | `detach.default`               |
|                | `zeros_like.default`           |
|                | `_foreach_mul_.Scalar`         |
|                | `_foreach_addcmul_.Scalar`     |
|                | `_foreach_sqrt.default`        |
|                | `_foreach_add_.Scalar`         |
|                | `_foreach_addcdiv_.Scalar`     |
| sgd            | `detach.default`               |
|                | `_foreach_add_.List`           |
| sgd (momentum) | `detach.default`               |
|                | `clone.default`                |
|                | `_foreach_add_.List`           |
|                | `_foreach_add_.List`           |
|                | `_foreach_mul_.Scalar`        |


### Analysis of impact to view ops
**TL;DR** Previous and current code can propagate `StridedShard` by maintaining the same split_factor (which I don't think this is correct for all cases), as long as the tensor dim is shardable on the mesh. We need follow up fix the let view op propagate arbitrary combination of `StridedShard`.

I marked `expectedFailure` for `test_view_propagation_not_supported_yet` in this PR. weifengpy plans to work on a broader support.

### Analysis of impact to random ops
Our current `OffsetBasedRNGTracker` uses the Shard placement to compute the shard idx of each shard the shard size. This will be used to decide the RNG offset. However, the offset computation didn't consider the split_factor in the `StridedShard`.

This PR didn't fix any offset computation based on `StridedShard`. It only allows the `StridedShard` to be treated equally as Shard, which is the same behavior before this PR when `StridedShard` is a subclass of Shard.


---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci 

[ghstack-poisoned]
